packages/analytics/amplify_analytics_pinpoint_dart/lib/src/impl/analytics_client/event_client/event_client.dart

```diff
@@ -215,21 +229,26 @@ class EventClient implements Closeable {
   }
 
   // If exception is recoverable, do not delete eventIds from local cache
-  void _handleRecoverableException(
-    SmithyHttpException e,
-    Map<String, StoredEvent> eventIdsToDelete,
-  ) {
+  void _logRecoverableException(Exception e) {
     _logger
       ..warn('putEvents - exception encountered: ', e)
       ..warn('Recoverable issue, will attempt to resend event batch');
-    eventIdsToDelete.clear();
   }
 
-  bool _isRetryable(String? message) {
-    return message != null &&
-        !_equalsIgnoreCase(message, 'ValidationException') &&
-        !_equalsIgnoreCase(message, 'SerializationException') &&
-        !_equalsIgnoreCase(message, 'BadRequestException');
+  bool _isRetryable(int statusCode) {
+    // Pinpoint service policy is for 500-599 status codes to be retryable
+    return statusCode >= 500 && statusCode < 600;
+  }
+
+  bool _shouldRetryEvent(String eventId) {
+    final timesFailed = _numFailuresByEvent[eventId] ?? 0;
+    if (timesFailed >= 3) {
+      _logger.warn('Retry limit exceeded, deleting event: $eventId');
+      _numFailuresByEvent.remove(eventId);
+      return false;
+    }
+    _numFailuresByEvent[eventId] = timesFailed + 1;
```

Comment:
question: If a retry is successful, will the IDs for that event stay in `_numFailuresByEvent` creating a memory leak? I didn't see where this is cleared when successful, but maybe I'm missing something.

As a larger question, why is it necessary to track which events need retry? If the failure is by HTTP status code, can't you assume all the events in a request need a retry and just retry the request itself? something like https://pub.dev/packages/retry?

Also, will the retry here be immediate (seemed so from testing), would we want some kind of backoff for the retry?

Reply:
Good call yes it would have, added code to fix that. 

Events can fail individually.  The underlying Pinpoint service sends separate responses for each event.  

Events send every 10 seconds.  There is no exponential backoff here, not sure if that's necessary since we always send events and can just retry on the next round.

Reply:
Makes sense, thanks for explanation.